File: BinarySearch.java
class BinarySearch {
    public int search(int[] nums, int target) {
        if(nums.length>1){
        int mid = (nums.length/2)-1;
        System.out.println(mid);
        int old_mid = mid;
        int rotate=0;
        //Checking for rotation in the array
        if((nums[mid]>=nums[0])&&(nums[mid]<=nums[nums.length-1])){
            System.out.println("There is no rotation");
            rotate=0;
        }
        else if((nums[mid]>=nums[0])&&(nums[mid]>=nums[nums.length-1])){
            System.out.println("There is a rotation in second half");
            rotate=2;
        }
         else if((nums[mid]<=nums[0])&&(nums[mid]<=nums[nums.length-1])){
             System.out.println("There is rotation in first half");
             rotate=1;
         }
        else return mid;
        if(rotate==1){
             System.out.println("Entering the half 1");
            return binarySearch(nums,mid,0,target, old_mid);
        }
        else if(rotate==2){
            System.out.println("Entering the half 2");
            return binarySearch(nums,nums.length-1,mid+1,target, old_mid);
        }
        else if(rotate==0){
            return binarySearch(nums,nums.length-1,0,target, old_mid);
        }
            return -1;
        }
        else if ((nums.length==1)&&(nums[0]==target)) {  return 0;}
        return -1;
    }
    public int binarySearch(int[] nums, int high, int low, int target, int old_mid){
        System.out.println(low);
        System.out.println(high);
        if(nums[old_mid]==target) {System.out.println("equal to old mid"+old_mid); return old_mid;}
        if(high>=low){
        int mid = low + (high-low)/2;
            System.out.println("mid is"+mid);
        if(nums[mid]>target) {//loop=1; 
            System.out.println("dividing half 1");
            high=mid-1; return binarySearch(nums,high,low,target,old_mid);}
        else if(nums[mid]<target) {//loop=2;
            System.out.println("dividing half 2");
            low=mid+1; return binarySearch(nums,high,low,target,old_mid);}
        else if(nums[mid]==target) return mid; 
        else return -1;
        }
        else return -1;
    }
}

File: 34_Find_First_and_LastPosition_of_Element_in_Sorted_Array/binary_search.cpp
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : I didnt realise that we should check if the current element is first/last repetition hence went in infinite loop


// Your code here along with comments explaining your approach in three sentences only
// Technique: Binary search
// As this is a sorted array and we have to find an element, BS is the most natural.
// Now we divide the problem in finding first position and finding the last position.
// Consider finding first: When we encounter target at mid, then  either this is the first position or there are more target to left.
// So we check if this is first, if yes return mid else search in left half (excluding the mid).


#include<bits/stdc++.h>
using namespace std;

void printArr(vector<int> arr) {
    for(int a: arr) {
        cout<<a<<" ";
    }
    cout<<endl;
}

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int targetStPos = binarySearchStart(nums,target);
        if(targetStPos == -1) {
            return vector<int>{-1,-1};
        }
        int targetEndPos = binarySearchEnd(nums,target);
        return vector<int>{targetStPos,targetEndPos};
    }

private:
    int binarySearchStart(vector<int>& nums, int target) {
        int st = 0, end = nums.size()-1;
        while(st<=end) {
            int mid = st + (end-st)/2;
            if(nums[mid] == target && (mid == 0 || nums[mid-1] < nums[mid])) {
                return mid;
            }
            if(nums[mid] >= target) {
                end = mid-1;
            }
            else {
                st = mid+1;
            }
        }
        return -1;
    }

    int binarySearchEnd(vector<int>& nums, int target) {
        int st = 0, end = nums.size()-1;
        while(st<=end) {
            int mid = st + (end-st)/2;
            if(nums[mid] == target && (mid == nums.size()-1 || nums[mid+1] > nums[mid])) {
                return mid;
            }
            if(nums[mid] <= target) {
                st = mid+1;
            }
            else {
                end = mid-1;
            }
        }
        return -1;
    }
};

int main() {
    vector<int> nums{5,7,7,8,8,10};
    Solution S;
    printArr(S.searchRange(nums, 6));
}